overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
 	return filePerms, nil
 }
 
-var allowablePathPatternRegexp = regexp.MustCompile(`^(/|(/[^/*{}]+)*(/\*|(/\*\*)?(/\*\.[^/*{}]+)?)?)$`)
+var (
+	// The following must be escaped as literals in a path pattern:
+	problematicChars = `\[\]`
+	// A single safe or escaped unsafe char in a path
+	safePathChar = fmt.Sprintf(`([^%s]|\\[%s])`, problematicChars, problematicChars)
+
+	// The following matches valid path patterns
+	allowablePathPattern       = fmt.Sprintf(`^/%s*$`, safePathChar)
+	allowablePathPatternRegexp = regexp.MustCompile(allowablePathPattern)
+)
+
+// Expands all groups in the given path pattern. Groups are enclosed by '{' '}'.
+// Returns a list of all the expanded path patterns, or an error if the given
+// pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	patternLen := len(pattern)
+	expanded := defaultPrefixes
+	currLiteralStart := 0
+	for i := 0; i < patternLen; i++ {
+		if pattern[i] == '\\' {
+			i += 1
+			if i == patternLen {
+				return nil, fmt.Errorf(`invalid path pattern: trailing unescaped '\' character: %q`, pattern)
+			}
+			continue
+		}
+		if pattern[i] == '}' {
+			return nil, fmt.Errorf(`invalid path pattern: unmatched '}' character: %q`, pattern)
+		}
+		if pattern[i] != '{' {
+			continue
+		}
+		infix := pattern[currLiteralStart:i]
```

Comment:
I'd probably move this definition a little later as it isn't needed just yet here :P